From 6d230b30fa4d4974644a208732ce308616181cb9 Mon Sep 17 00:00:00 2001
From: Brian Delwiche <delwiche@google.com>
Date: Wed, 15 Jan 2025 23:36:59 +0000
Subject: [PATCH] Fix authentication bypass bug in SMP

It is possible for a malicious peer to bypass SMP authentication by
claiming to have OOB data and proceeding with pairing, exploiting the
fact that the Android stack sets the OOB randomizer to zero if no local
OOB data is available.

Drop the connection if a peer claims it has OOB data but no local OOB
data has been stored.

Bug: 251514171
Test: m com.android.btservices
Ignore-AOSP-First: security
Tag: #security
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:faea50382d2b1932abac40b76507d9bcd374635e)
Merged-In: I85a121aea7b37318f8cfbcdb8075ece67806957e
Change-Id: I85a121aea7b37318f8cfbcdb8075ece67806957e
---
 system/stack/smp/smp_act.cc | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/system/stack/smp/smp_act.cc b/system/stack/smp/smp_act.cc
index 35e2643319..ace755435f 100644
--- a/system/stack/smp/smp_act.cc
+++ b/system/stack/smp/smp_act.cc
@@ -1966,6 +1966,17 @@ void smp_process_secure_connection_oob_data(tSMP_CB* p_cb, tSMP_INT_DATA* /* p_d
     }
   }
 
+  if (p_cb->peer_oob_flag == SMP_OOB_PRESENT &&
+      !p_sc_oob_data->loc_oob_data.present) {
+    log::warn(
+        "local OOB data is not present but peer claims to have received it; dropping "
+        "connection");
+    tSMP_INT_DATA smp_int_data{};
+    smp_int_data.status = SMP_OOB_FAIL;
+    smp_sm_event(p_cb, SMP_AUTH_CMPL_EVT, &smp_int_data);
+    return;
+  }
+
   if (!p_sc_oob_data->peer_oob_data.present) {
     log::verbose("peer OOB data is absent");
     p_cb->peer_random = {0};
-- 
2.34.1

